### PR TITLE
Fix/pob/ib 572

### PR DESCRIPTION
**Pull request by liangjn16**

### CODE REVIEW

File: consensus/pob/pob.go [hunk 1]
```diff
@@ -475,10 +477,8 @@ func (p *PoB) addExistingBlock(blk *block.Block, parentBlock *block.Block, repla
 }
 
 func (p *PoB) updateInfo(node *blockcache.BlockCacheNode) {
-	updateWaterMark(node)
 	updateLib(node, p.blockCache)
-	staticProperty.updateWitness(p.blockCache.LinkedRoot().Active())
-	if staticProperty.isWitness(p.account.ReadablePubkey()) {
+	if staticProperty.isWitness(p.account.ReadablePubkey(), p.blockCache.LinkedRoot().Pending()) {
```
Comment (x-zho14) [inline consensus/pob/pob.go]: don't know why

File: consensus/pob/pob.go [hunk 2]
```diff
@@ -95,7 +95,7 @@ func New(account *account.KeyPair, baseVariable global.BaseVariable, blockCache
 		mu:               new(sync.RWMutex),
 	}
 	continuousNum = baseVariable.Continuous()
-	staticProperty = newStaticProperty(p.account, blockCache.LinkedRoot().Active())
+	staticProperty = newStaticProperty(p.account, blockCache.LinkedRoot().PendingNum())
```
Comment (ziranliu) [inline consensus/pob/pob.go]: delete PendingNum()

File: consensus/pob/pob.go [hunk 3]
```diff
@@ -338,7 +340,7 @@ func (p *PoB) scheduleLoop() {
 					select {
 					case <-generateBlockTicker.C:
 					}
-					if witnessOfNanoSec(time.Now().UnixNano()) != pubkey {
+					if witnessOfNanoSec(t.UnixNano(), witnessList) != pubkey {
```
Comment (ziranliu) [inline consensus/pob/pob.go]: should be time.Now().

File: core/blockcache/block_cache.go [hunk 4]
```diff
@@ -170,6 +172,34 @@ func NewVirtualBCN(parent *BlockCacheNode, blk *block.Block) *BlockCacheNode {
 	return bcn
 }
 
+func (bcn *BlockCacheNode) updateVaildWitness(parent *BlockCacheNode, witness string) {
```
Comment (x-zho14) [inline core/blockcache/block_cache.go]: list->map

File: core/blockcache/block_cache.go [hunk 5]
```diff
@@ -170,6 +172,34 @@ func NewVirtualBCN(parent *BlockCacheNode, blk *block.Block) *BlockCacheNode {
...
+			witness = ""
+		}
+	}
+	if witness != "" {
+		bcn.VaildWitness = append(bcn.VaildWitness, witness)
+	}
+}
+
+func (bcn *BlockCacheNode) removeVaildWitness(root *BlockCacheNode) {
+	if &bcn.Active()[0] != &root.Pending()[0] || strings.Compare(bcn.Head.Witness, root.Head.Witness) == 0 {
```
Comment (ziranliu) [inline core/blockcache/block_cache.go]: DON'T!  So trick and confusing!

File: consensus/pob/algorithm.go [hunk 6]
```diff
@@ -136,43 +136,35 @@ func verifyBlock(blk *block.Block, parent *block.Block, lib *block.Block, txPool
...
-	confirmedNode := calculateConfirm(node, bc.LinkedRoot())
-	if confirmedNode != nil {
-		bc.Flush(confirmedNode)
-		metricsConfirmedLength.Set(float64(confirmedNode.Head.Number+1), nil)
-	}
-}
+	confirmLimit := int(staticProperty.NumberOfWitnesses*2/3 + 1)
+	root := bc.LinkedRoot()
+	if len(node.VaildWitness) >= confirmLimit {
+		if &node.Active()[0] == &bc.LinkedRoot().Pending()[0] {
```
Comment (ziranliu) [inline consensus/pob/algorithm.go]: DON'T!

File: verifier/block_base_tx.go [hunk 7]
```diff
@@ -29,12 +30,16 @@ func NewBaseTx(blk *block.Block, parent *block.Block) (*tx.Tx, error) {
 	return tx, nil
 }
 
-func baseTxData(b *block.Block, pb *block.Block) (string, error) {
+func baseTxData(b, pb *block.Block, witnessList *blockcache.WitnessList) (string, error) {
 	if pb != nil {
-		return fmt.Sprintf(`[{"parent":["%v", "%v"]}]`, pb.Head.Witness, pb.CalculateGasUsage()), nil
+		witnessChanged := false
+		if witnessList != nil && &witnessList.Active()[0] != &witnessList.Pending()[0] {
```
Comment (ziranliu) [inline verifier/block_base_tx.go]: DON'T!

File: consensus/pob/pob.go [hunk 8]
```diff
@@ -315,21 +315,28 @@ func (p *PoB) blockLoop() {
...
+	tGenBlock := time.NewTicker(20 * time.Millisecond)
 	tMetricsMode := time.NewTicker(3 * time.Second)
 	defer tGenBlock.Stop()
 	defer tMetricsMode.Stop()
 
+	var slotFlag int64
 	for {
 		select {
 		case <-tGenBlock.C:
+			// Don't delete,avoid time error
```
Comment (x-zho14) [inline consensus/pob/pob.go]: Don't delete, avoid time error.

Comment (x-zho14) [inline consensus/pob/pob.go]: Don't delete in case of time error.

File: consensus/pob/pob.go [hunk 9]
```diff
@@ -315,21 +315,28 @@ func (p *PoB) blockLoop() {
...
 			t := time.Now()
+			_, head := p.txPool.PendingTx()
+			witnessList := head.Active()
 			pubkey := p.account.ReadablePubkey()
-			if witnessOfNanoSec(t.UnixNano()) == pubkey && !staticProperty.SlotUsed[slotOfSec(t.Unix())] && p.baseVariable.Mode() == global.ModeNormal {
-				staticProperty.SlotUsed[slotOfSec(t.Unix())] = true
+			if witnessOfNanoSec(t.UnixNano(), witnessList) == pubkey && slotFlag != slotOfSec(t.Unix()) && p.baseVariable.Mode() == global.ModeNormal {
+				slotFlag = slotOfSec(t.Unix())
+				p.quitGenerateMode = make(chan struct{})
+				staticProperty.SlotUsed[t.Unix()] = true // never delete
```
Comment (x-zho14) [inline consensus/pob/pob.go]: delete this line?

File: core/blockcache/block_cache.go [hunk 10]
```diff
@@ -294,7 +324,6 @@ func NewBlockCache(baseVariable global.BaseVariable) (*BlockCacheImpl, error) {
 	if err := bc.updatePending(bc.linkedRoot); err != nil {
 		return nil, err
 	}
-	bc.linkedRoot.LibWitnessHandle()
```
Comment (x-zho14) [inline core/blockcache/block_cache.go]: don't know why

File: core/blockcache/block_cache.go [hunk 11]
```diff
@@ -370,9 +399,7 @@ func (bc *BlockCacheImpl) applyLink(b []byte, p conAlgo) (err error) {
 func (bc *BlockCacheImpl) applySetRoot(b []byte) (err error) {
 	bcn, bo := bc.hmget(b)
 	if bo {
-		bc.flush(bcn)
-		bc.delSingle()
-		bc.updateLongest()
+		bc.Flush(bcn)
```
Comment (x-zho14) [inline core/blockcache/block_cache.go]: don't know why

**Pull request merged**